keps/3122-expose-flavors-in-localqueue-status/README.md

```diff
@@ -0,0 +1,158 @@
+# KEP-3122: Expose Flavors in LocalQueue Status
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [User Stories (Optional)](#user-stories-optional)
+    - [Story 1](#story-1)
+  - [Notes/Constraints/Caveats (Optional)](#notesconstraintscaveats-optional)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [API](#api)
+  - [Implementation overview](#implementation-overview)
+  - [Test Plan](#test-plan)
+      - [Prerequisite testing updates](#prerequisite-testing-updates)
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Summary
+
+This KEP introduces a new status field in LocalQueue, allowing users to see
+all currently available ResourceFlavors in the LocalQueue.
+
+## Motivation
+
+Currently, users without RBAC access to ResourceFlavors cannot view the list
+of available flavors. Depending on the RBAC rules, users might also lack read
+access to ClusterQueues. Providing users with information about the available
+flavors is useful, as it gives them an idea of the capabilities provided by
+a LocalQueue (e.g., a flavor might include newer GPUs).
+
+### Goals
+
+- Provide a possibility to see all currently available ResourceFlavors in 
+  the LocalQueue.
+
+### Non-Goals
+
+- Verify that the ResourceFlavors exist and show only the existing flavors.
+
+## Proposal
+
+Introduce a new status field `availableFlavors` in LocalQueue 
+that will be updated when ClusterQueue flavors are modified.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user I want to see the list of all ResourceFlavors available in each LocalQueue 
+due to the RBAC configuration for ClusterQueue in my cluster I cannot inspect the 
+ClusterQueue objects directly, only LocalQueues.
+
+### Notes/Constraints/Caveats (Optional)
+
+### Risks and Mitigations
+
+## Design Details
+
+### API
+
+Create `AvailableFlavor` API object:
+
+```go
+type AvailableFlavor struct {
+  // name of the flavor.
+  Name ResourceFlavorReference `json:"name"`
+  // resources used in the flavor.
+  // +listType=set
+  // +optional
+  Resources []string `json:"resources,omitempty"`
```

Comment:
```suggestion
  // resourceGroups used in the flavor.
  // +listType=set
  // +optional
  ResourceGroups []string `json:"resourceGroups,omitempty"`
```

Isn't this "resources" semantic bring admins new concept, and this will increase cognitive load?
Couldn't we keep using the "resourceGroup" between clusterQueue and localQueue?

If we use "resources" here, the new `cq.resourceGroups:lq.resources = 1:1` semantic will be brought.
Additionally, I would like to avoid using the "resources" as a field or concept name since we have already been "resources" in some places, and the additional "resources" concept will bring us more complexity.

cc: @mimowo @alculquicondor

Reply:
I think "ResourceGoups" as string could be confusing as well [type](https://github.com/kubernetes-sigs/kueue/blob/a7681277ea932aef0b233e635447c0e16b31e116/apis/kueue/v1alpha1/cohort_types.go#L62) associated with it, so using it here are stings might be confusing as well.

I think the 1:1 mapping wouldn't be perfect anyway since we decided not to expose the quotas, which are part of ResourceGroups.

If we attempt to make 1:1 mapping I think it would look like:
```yaml
resourceGroups:
- coveredResources: [cpu, memory]
  flavors:
  - name: on-demand
    resources: [cpu, memory] # without quotas
  - name: spot
    resources: [cpu, memory] # without quotas
- coveredResources: [some-resource]
  flavors:
  - name: another
    resources: [some-resource]  # without quotas
```
vs (as currently proposed):
```yaml
availableflavors:
  - name: on-demand
    resources: [cpu, memory]
  - name: spot
    resources: [cpu, memory]
  - name: another
    resources: [some-resource]
```
TBH the first version looks quite verbose as for users.

Reply:
Sorry. I wanted to indicate `coveredResource` instead of `resourceGroups` as follows:

```yaml
availableFlavors:
- name: on-demand
  coveredResources: ["cpu", "memory"]
- name: spot
  coveredResources: ["cpu", "memory"]
```

But, it seems that the below your proposal looks more reasonable to me.
Thanks!

> ```yaml
> availableflavors:
>   - name: on-demand
>     resources: [cpu, memory]
>   - name: spot
>     resources: [cpu, memory]
>   - name: another
>     resources: [some-resource]
> ```

Although, this may be another enhancement, we may want to expose `coveredResources`, `nodeTaints`, and `nodeLabels` since I think what users want to know is available resource names and how to use the resourceFlavors. 

Actually, the batch users can not access the resourceFlavor information and then, can not directly specify the resourceFlavor.

The `availableCoveredResources` allows batch users to understand which resources can be available for the localQueue, and the `nodeLabels` and `nodeTaints` allow the batch users to know which nodeSelectors and tolerations must be specified if they want to use the flavors.

```yaml
status:
  availableCoveredResources: ["cpu", "memory"]
  availableFlavors:
  - name: on-demand
    resources: ["cpu", "memory"]
    nodeLabels:
      instance-type: on-demand
  - name: spot
    resources: ["cpu", "memory"]
    nodeTaints:
    - effect: NoSchedule
       key: spot
       value: "true"
```

@mimowo WDYT?

Reply:
I think all of the suggested fields `availableCoveredResources`, `nodeLabels`, and `nodeTaints` might be useful, but I want to also balance the gain / effort ratio.

Let me check with @KPostOffice and @varshaprasad96 on the usefulness of those.

Reply:
If we don't have a clear need to introduce them in the first iteration, then I would suggest to add a graduation point for GA:
"Re-evaluate adding the availableCoveredResources, nodeLabels, and nodeTaints fields based on user-feedback".

Reply:
I agree with you. At least, I believe that we need to expose the `nodeTaints` since batch users need to specify the the tolerations based on the ResourceFlavor taint feature: https://kueue.sigs.k8s.io/docs/concepts/resource_flavor/#resourceflavor-taints

If the nodeTaints are not exposed to users, they can not use the resourceFlavor.

Reply:
> In that case, I'm ok to include nodeTaints and nodeLabels in the first iteration, but for availableCoveredResources I think we can wait until feedback. WDYT?

SGTM.
I also would recommend adding the `availableCoveredResources` to future work or an alternative section.

Reply:
@mbobrovskyi please update the KEP accordingly:
- extend AvailableFlavor API by the `resources`, `nodeTaints` and `nodeLabels` fields
- add graduation criteria for GA to re-evaluate adding `availableCoveredResources`

Reply:
Sorry, what's the difference between `resources` and `availableCoveredResources`?

Reply:
Aha, and `availableCoveredResources` would union across resource groups, whilst `resources` would be per flavor.

Reply:
> I think technically the lists could differ

No, they don't. There is validation that makes sure they are always the same as the list of resources in a flavor. `coveredResources` was only added to make it more visually obvious about which resources are associated to each resource group.

Reply:
Oh right, I missed that. In that case I see no point having `availableCoveredResources` at the LQ status level.

Reply:
@alculquicondor @mimowo My motivations for the "availableCoveredResources" was to improve the visibility since when they use multiple flavors, resources, an transformation, the `availableFlavors` could be complexed like this:

```yaml
availableFlavors:
- name: vender-a
  resources: ["cpu", "memory", "vender-a/xxx"]
- name: vender-b
  resources: ["cpu", "memory", "vender-b/yyy", "vender-c/zzz"]
- name: vender-c
  resources: ["cpu", "memory", "vender-c/gpu.5gb", "vender-c/gpu.10gi", "vender-c/gpu.20gi", "vender-c/gpu.40gi", "vender-c/vf.0", "vender-c/vf.1", "vender-c/vf.2", "vender-c/vf.3", "vender-c/vf.4", "vender-c/vf.5", "vender-c/vf.6", "vender-c/vf.7", "ephemeral-storage"]
...
```

Actually, the number of resources often larger when we use the special devices like SRIO-V VirtualFunctions (VFs) and FPGA and so on since they are often created each GPU.

(Indeed, I have and maintain ClusterQueues contained many many devices in production)

But, I do not offer to contain the `availableCoveredResources` in the first iteration since this is trade-off between complexity and visibility. After collecting user feedback, we can revisit here.

Reply:
> add graduation criteria for GA to re-evaluate adding availableCoveredResources

Shouldn't it be on the separate KEP?

Reply:
@mbobrovskyi Could you mention the availableCoveredResources in the alternative or future work section?
The field could improve visibility in a situation where the CQ has huge resource names.

Reply:
Done